Source/WebCore/page/PointerLockController.cpp

```diff
@@ -40,44 +42,83 @@
 
 namespace WebCore {
 
+class UnadjustedMovementPlatformMouseEvent : public PlatformMouseEvent {
+public:
+    UnadjustedMovementPlatformMouseEvent(const PlatformMouseEvent &src)
+        : PlatformMouseEvent(src.position(), src.globalPosition(), src.button(), src.type(), src.clickCount(), src.modifiers(), src.timestamp(), src.force(), src.syntheticClickType(), src.pointerId())
+    {
+        m_movementDelta = m_unadjustedMovementDelta = src.unadjustedMovementDelta();
+    }
+};
+
 PointerLockController::PointerLockController(Page& page)
     : m_page(page)
 {
 }
 
-void PointerLockController::requestPointerLock(Element* target)
+PointerLockController::~PointerLockController()
+{
+}
```

Comment:
Sorry I don't understand -- why do we need this definition?

Reply:
DeferredPromise is a forward declaration in PointerLockController.h. So when ~PointerLockController() is synthesized by the compiler, it doesn't know how to destruct DeferredPromise without its actual declaration from DeferredPromise.h. You'd get this error if we eliminated it:

```
In file included from /Users/jahoward/git/ext/WebKit/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource263.cpp:1:
In file included from /Users/jahoward/git/ext/WebKit/Source/WebCore/WebCorePrefix.h:165:
In file included from /Users/jahoward/git/ext/WebKit/WebKitBuild/Debug/usr/local/include/wtf/HashMap.h:25:
In file included from /Users/jahoward/git/ext/WebKit/WebKitBuild/Debug/usr/local/include/wtf/HashTable.h:33:
In file included from /Users/jahoward/git/ext/WebKit/WebKitBuild/Debug/usr/local/include/wtf/HashTraits.h:26:
In file included from /Users/jahoward/git/ext/WebKit/WebKitBuild/Debug/usr/local/include/wtf/HashFunctions.h:26:
/Users/jahoward/git/ext/WebKit/WebKitBuild/Debug/usr/local/include/wtf/RefPtr.h:43:16: error: member access into incomplete type 'WebCore::DeferredPromise'
            ptr->deref();
```

As I see it, there are two options to fix: We can #include "DeferredPromise.h" in "PointerLockController.h" or we can implement the destructor in the implementation file which has access to the actual declaration of DeferredPromise. I chose the latter since I prefer to forward declare where possible.

Reply:
Thanks for the explanation, and I agree with your solution